pilot/pkg/config/kube/crd/controller/controller.go

```diff
@@ -182,27 +182,39 @@ func (c *controller) createInformer(
 	handler := &kube.ChainHandler{}
 	handler.Append(c.notify)
 
+	// Wrap the list function with supplied validation function, so that invalid CRD does not enter store.
+	vlf := func(opts meta_v1.ListOptions) (result runtime.Object, err error) {
```

Comment:
So i think this is redundant

Reply:
See my response here https://github.com/istio/istio/pull/19172#discussion_r350063859

Reply:
Which version do you use?

Reply:
Master

Reply:
if you are asking k8s, it is `1.14.6` in the env in which this happened.

Reply:
Do you test with this pr? Can it prevent panic?

BTW, Can you show me your invalid VS, i want to do a dig to see what's wrong

Reply:
Here is  the one that caused Panic. Please see that `destination` field is missing. I did not test with this PR yet. It would be great if you can reproduce and dig deep.

```
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  clusterName: ""
  creationTimestamp: "2019-11-20T23:49:18Z"
  generation: 1
  name: asr-bg-vs
  namespace: slb
  resourceVersion: "2818162890"
  selfLink: /apis/networking.istio.io/v1alpha3/namespaces/slb/virtualservices/asr-bg-vs
  uid: 5e0f52f7-0bf0-11ea-ace9-14187760f413
spec:
  gateways:
  - testns/test-gateway
  hosts:
  - host.net
  http:
  - route:
    - host: wikipedia.org
    timeout: 5s
```

Reply:
I doubt this panic is because `w.Destination` = nil

Reply:
It is. But validation should have caught it. See the test cases I added in this PR for validating that - Validation actually catches that. That is why I am doubting, it is not validated.

Reply:
Validate can't catch it, you can add a debug log in the validation function, it will verify it.

Reply:
Please see the new test case "with no destination" that I added - It seems to catch if destination does not exist. Am I missing some thing?

Reply:
I verified:

```


2019-11-25T11:59:49.701881Z	error	failed to add CRD. New value: &{{ } {asr-bg-vs  default /apis/networking.istio.io/v1alpha3/namespaces/default/virtualservices/asr-bg-vs 52259044-6ed1-42fd-b863-4737ba6e2c37 2539 1 2019-11-25 11:59:49 +0000 UTC <nil> <nil> map[] map[kubectl.kubernetes.io/last-applied-configuration:{"apiVersion":"networking.istio.io/v1alpha3","kind":"VirtualService","metadata":{"annotations":{},"name":"asr-bg-vs","namespace":"default"},"spec":{"gateways":["testns/test-gateway"],"hosts":["host.net"],"http":[{"route":[{"host":"wikipedia.org"}],"timeout":"5s"}]}}
] [] []  []} map[gateways:[testns/test-gateway] hosts:[host.net] http:[map[route:[map[host:wikipedia.org]] timeout:5s]]]}, error: failed to validate CRD &{{virtual-service networking.istio.io v1alpha3 asr-bg-vs default cluster.local map[] map[kubectl.kubernetes.io/last-applied-configuration:{"apiVersion":"networking.istio.io/v1alpha3","kind":"VirtualService","metadata":{"annotations":{},"name":"asr-bg-vs","namespace":"default"},"spec":{"gateways":["testns/test-gateway"],"hosts":["host.net"],"http":[{"route":[{"host":"wikipedia.org"}],"timeout":"5s"}]}}
] 2539 2019-11-25 11:59:49 +0000 UTC} hosts:"host.net" gateways:"testns/test-gateway" http:<route:<> timeout:<seconds:5 > > }, error: 1 error occurred:
	* destination is required

```

Reply:
The validation indeed runs

Reply:
btw 
```
# istioctl version
client version: 1.5-alpha.3b7b43f911ab6c69bf0da9623966521563cff3f8
control plane version: 1.5-alpha.3b7b43f911ab6c69bf0da9623966521563cff3f8
data plane version: 1.5-alpha.3b7b43f911ab6c69bf0da9623966521563cff3f8 (2 proxies)

```

Reply:
@hzxuzhonghu This makes me think, under some condition, the invalid VS is making in to cache store (by passing validation), could be a bug in client go or some thing else. 
Not sure what it is - do you think adding validation to List, makes it more robust?

BTW, when the validation failed in your case, what happens if we trigger manual push/restart pilot? - I am guessing it will work fine?

Reply:
>do you think adding validation to List, makes it more robust?

It will not, just a duplicate.

>BTW, when the validation failed in your case, what happens if we trigger manual push/restart pilot? - I am guessing it will work fine?

Wow, this is an interesting point, i missed it. If a trigger from other events, the invalid configs will be read by pilot, and the initContext will be called, and then panic.

Reply:
> the invalid configs will be read by pilot

Where will it read from? 

My point about adding validation to List is here https://github.com/istio/istio/blob/master/pilot/pkg/config/kube/crd/controller/controller.go#L325 - not in the `ListFunc`. Specifically [here](https://github.com/istio/istio/blob/master/pilot/pkg/config/kube/crd/controller/controller.go#L351) - because that is what is fed to Push Context? But I guess that has performance minor perf impact - I am guessing the perf is not significant because, the object is already converted (which is the costly part) and validation just have to run through. I think it would be useful for folks not using galley. If performance is a major concern, probably we can guard it with a feature flag - so folks who do not use galley, can set it to true?  WDYT?

Reply:
SGTM

Reply:
Same applies to https://github.com/istio/istio/blob/master/pilot/pkg/config/kube/crd/controller/controller.go#L283

Reply:
Great. I will change. Should I guard it with flag?

Reply:
Yes, let's guard it with flag.

Reply:
Implemented it with gaurd. PTAL